src/control_devices/CCPACSControlSurfaceBorderLeadingEdge.cpp

```diff
@@ -0,0 +1,201 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "CCPACSControlSurfaceBorderLeadingEdge.h"
+#include "CCPACSControlSurfaceOuterShapeLeadingEdge.h"
+#include "CCPACSLeadingEdgeDevice.h"
+#include "CControlSurfaceBorderBuilder.h"
+#include "tigletaxsifunctions.h"
+
+#include "CNamedShape.h"
+#include "Debugging.h"
+#include <cassert>
+
+namespace tigl
+{
+
+CCPACSControlSurfaceBorderLeadingEdge::CCPACSControlSurfaceBorderLeadingEdge(
+    CCPACSControlSurfaceOuterShapeLeadingEdge* parent, CTiglUIDManager* uidMgr)
+    : generated::CPACSControlSurfaceBorderLeadingEdge(parent, uidMgr)
+{
+}
+
+TopoDS_Wire CCPACSControlSurfaceBorderLeadingEdge::GetWire(PNamedShape wingShape, gp_Vec upDir) const
+{
+    assert(wingShape);
+
+    // Compute cutout plane
+    CTiglControlSurfaceBorderCoordinateSystem coords = GetCoordinateSystem(upDir);
+    CControlSurfaceBorderBuilder builder(coords, wingShape->Shape());
+
+    TopoDS_Wire wire;
+    if (getXsiTE() > 1 - 1e-3) {
+
+        wire = builder.wholeWingBorder();
+    }
+
+    else if (GetLeadingEdgeShape_choice2().is_initialized()) {
+        throw CTiglError("Leading edge device with LeadingEdgeShape not implemented");
+    }
+
+    else if (GetInnerShape_choice1().is_initialized()) {
+
+        double xsiTEUpper;
+        double xsiTELower;
+
+        if (GetXsiTE_choice1().is_initialized()) {
+            xsiTEUpper = GetXsiTE_choice1()->GetXsi();
+            xsiTELower = GetXsiTE_choice1()->GetXsi();
+        }
+        else {
+            if (!GetXsiTEUpper_choice2().is_initialized() || !GetXsiTELower_choice2().is_initialized()) {
+                throw CTiglError("LED relative chordwise trailing edge xsiTE and/ or xsiTEUpper and xsiTELower not "
+                                 "sepcified in CPACS file.");
+            }
+            xsiTEUpper = GetXsiTEUpper_choice2().value();
+            xsiTELower = GetXsiTELower_choice2().value();
+        }
+
+        wire = builder.borderWithInnerShape(GetInnerShape_choice1()->GetRelHeightTE(),
+                                            GetInnerShape_choice1()->GetXsiTE(), xsiTEUpper, xsiTELower);
+    }
+
+    else if (GetAirfoil_choice3().is_initialized()) {
+
+        wire = GetAirfoilWire(coords);
+    }
+
+    else {
+        throw CTiglError("Airfoil definition of leading edge device innner/outer border missing.");
+    }
+
+#ifdef DEBUG
+    DEBUG_SCOPE(debug);
+    debug.dumpShape(wire, "controldevice-border-wire");
+#endif
+
+    return wire;
+}
+
+TopoDS_Wire
+CCPACSControlSurfaceBorderLeadingEdge::GetAirfoilWire(CTiglControlSurfaceBorderCoordinateSystem& coords) const
```

Comment:
Suggestion for more DRYness

This function really only uses a CCPACSWingProfile. So most of the implemenation can move to a function that is shared in the implementation for `CCPACSControlSurfaceBorderLeadingEdge` and `CCPACSControlSurfaceBorderTrailingEdge`.